### PR TITLE
router: add ability to select cluster via request header

### PR DESCRIPTION
**Pull request by mattklein123**
fixes https://github.com/lyft/envoy/issues/308

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @rshriram @GregHanson

**Comment by junr03** [comment]
@timperrett

File: docs/configuration/http_conn_man/route_config/route.rst [hunk 1]
```diff
@@ -40,15 +41,28 @@ path
...
 cluster
   *(sometimes required, string)* If the route is not a redirect (*host_redirect* and/or
-  *path_redirect* is not specified), one of *cluster* or *weighted_clusters* must be specified.
-  When a *cluster* is specified, its value indicates the upstream cluster to which the request
-  should be forwarded to.
+  *path_redirect* is not specified), one of *cluster*, *cluster_header*, or *weighted_clusters* must
+  be specified. When *cluster* is specified, its value indicates the upstream cluster to which the
+  request should be forwarded to.
+
+cluster_header
```
Comment (timperrett) [inline docs/configuration/http_conn_man/route_config/route.rst]: Awesome :-) Just so im 100% clear, this is still going to work if the clusters are not known a-priori? (that is, they are discovered via CDS?)

Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route.rst]: Yes, clusters do not need to be defined ahead of time.

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -448,6 +448,7 @@ const std::string Json::Schema::ROUTE_ENTRY_CONFIGURATION_SCHEMA(R"EOF(
       "prefix" : {"type" : "string"},
       "path" : {"type" : "string"},
       "cluster" : {"type" : "string"},
+      "cluster_header" : {"type" : "string"},
       "weighted_clusters": {"$ref" : "#/definitions/weighted_clusters"},
       "host_redirect" : {"type" : "string"},
```
Comment (rshriram) [inline source/common/json/config_schemas.cc]: I suggest a one of here. cluster,cluster_header,weighted_cluster. That allows you to get rid of the checks in config_impl.cc

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -91,16 +92,20 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost, const Json:
...
   // Check to make sure that we are either a redirect route or we have a cluster.
   if (!(isRedirect() ^ have_cluster)) {
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
 
   if (have_cluster) {
-    if (!cluster_name_.empty() && have_weighted_clusters) {
-      throw EnvoyException("routes must have either single or weighted_clusters as target");
+    // This is a trick to do a three-way XOR. It would be nice if we could do this with the JSON
+    // schema but there is no obvious way to do this.
```
Comment (rshriram) [inline source/common/router/config_impl.cc]: thats called oneOf. :)

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -261,26 +283,23 @@ const Route* RouteEntryImplBase::clusterEntry(uint64_t random_value) const {
...
   if (isRedirect()) {
     return;
   }
 
-  // We could have either a normal cluster or weighted cluster
-  if (weighted_clusters_.empty()) {
-    if (!cm.get(this->clusterName())) {
-      throw EnvoyException(fmt::format("route: unknown cluster '{}'", this->clusterName()));
+  if (!cluster_name_.empty()) {
+    if (!cm.get(cluster_name_)) {
```
Comment (rshriram) [inline source/common/router/config_impl.cc]: Doesnt this result in error if in config I specify routes with clusters, that do not yet exist (because the cds query has not yet been made)?

File: source/common/router/config_impl.cc [hunk 5]
```diff
@@ -261,26 +283,23 @@ const Route* RouteEntryImplBase::clusterEntry(uint64_t random_value) const {
...
+      throw EnvoyException(fmt::format("route: unknown cluster '{}'", cluster_name_));
     }
-  } else {
+  } else if (!weighted_clusters_.empty()) {
```
Comment (rshriram) [inline source/common/router/config_impl.cc]: same here. And I think this is what @GregHanson was referring to in gitter

File: source/common/router/config_impl.cc [hunk 6]
```diff
@@ -359,12 +377,17 @@ VirtualHostImpl::VirtualHostImpl(const Json::Object& virtual_host, Runtime::Load
   }
 
   for (const Json::ObjectPtr& route : virtual_host.getObjectArray("routes")) {
-    if (route->hasObject("prefix")) {
+    bool has_prefix = route->hasObject("prefix");
+    bool has_path = route->hasObject("path");
+    if (!(has_prefix ^ has_path)) {
+      throw EnvoyException("routes must specify either prefix or path");
+    }
```
Comment (rshriram) [inline source/common/router/config_impl.cc]: This can once again be solved by JsonSchema. with oneOf. @ccaraman thoughts?

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (mattklein123) [inline source/common/json/config_schemas.cc]: I already discussed this with @ccaraman. AFAICT, there is no way to use oneof in this context in a JSON schema (intermixed with properties inside an object). If you know otherwise let me know. Same comment applies to below.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (mattklein123) [inline source/common/router/config_impl.cc]: Yes, it will still error. My thinking on this was to leave the behavior as is for a statically defined routes. Theoretically I could see a case where someone might want to have a static route table, but then load the clusters from CDS, but this is an edge case. More common will be to use cluster_header. In the RDS case, I'm planning on disabling the checking so when a route table is loaded from RDS it will *not* assume that clusters already exist.

Basically, I would like to keep the checking functionality, at least behind an option, because it's useful for error checking static configs. IMO I would leave this as is for now and see how it looks when I do the RDS change next.

File: source/common/router/config_impl.cc (hunk 5 above)
Comment (mattklein123) [inline source/common/router/config_impl.cc]: same comment as above.

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (mattklein123) [inline source/common/router/config_impl.cc]: same comment as above.

File: source/common/router/config_impl.cc (hunk 6 above)
Comment (mattklein123) [inline source/common/router/config_impl.cc]: same comment as above.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: Maybe it's worthwhile to leave a comment, even if it is temporary? I was also a little puzzled by this when I was going over the PR last night.

Comment (mattklein123) [inline source/common/router/config_impl.cc]: Will add comment

**Comment by mattklein123** [comment]
@rshriram @GregHanson @junr03 added comment

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (ccaraman) [inline source/common/json/config_schemas.cc]: I prefer we keep the code checks in place. They are easy to follow. If we were to move this to schema check, it would be hard to read and a lot harder to validate a correct schema without adding more tests.
